bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -325,268 +224,72 @@ impl AuthenticationService {
         authentication_data: Arc<OidcAuthenticationData>,
         callback_url: String,
     ) -> Result<Arc<Client>, AuthenticationError> {
-        let Some(client) = self.client.read().unwrap().clone() else {
-            return Err(AuthenticationError::ClientMissing);
-        };
-
-        let oidc = client.inner.oidc();
-
-        let url =
+        let callback_url =
             Url::parse(&callback_url).map_err(|_| AuthenticationError::OidcCallbackUrlInvalid)?;
 
-        let response = AuthorizationResponse::parse_uri(&url)
-            .map_err(|_| AuthenticationError::OidcCallbackUrlInvalid)?;
-
-        let code = match response {
-            AuthorizationResponse::Success(code) => code,
-            AuthorizationResponse::Error(err) => {
-                if err.error.error == AccessDenied {
-                    // The user cancelled the login in the web view.
-                    return Err(AuthenticationError::OidcCancelled);
-                }
-                return Err(AuthenticationError::OidcError {
-                    message: err.error.error.to_string(),
-                });
-            }
-        };
-
-        if code.state != authentication_data.state {
-            return Err(AuthenticationError::OidcCallbackUrlInvalid);
-        };
-
-        RUNTIME.block_on(async move {
-            oidc.finish_authorization(code).await?;
-
-            oidc.finish_login()
-                .await
-                .map_err(|e| AuthenticationError::OidcError { message: e.to_string() })
-        })?;
-
-        let user_id = client.inner.user_id().unwrap().to_owned();
-        let session =
-            client.inner.oidc().full_session().ok_or(AuthenticationError::SessionMissing)?;
+        // Login and get the user ID that was resolved by the authentication server.
+        let client = RUNTIME.block_on(self.inner.login_with_oidc_callback(
+            OidcAuthorizationData {
```

Comment:
Looking at it, the `OidcAuthenticationData` struct can likely be removed now, in favour of exposing this one through uniffi.